Прохожу спринт 13

EntityVadim opened the pull request:
Ревью, прошу тебя внести комментарии по коду, может, где-то, что-то стоит поменять, улучшить и т.д. или ссылки на полезные статьи, как я могу изменить свой код, может какие-то лайфхаки крутые подскажешь.
Буду тебе очень сильно в этом благодарен!

EntityVadim committed:
- Создал файл с кодом общего дегодирования данных (Позже попытаюсь реализовать его)

EntityVadim committed:
- Устранил дубляж кода у себя

EntityVadim committed:
- Создал протоколы и презентер;
- Перенес логику запроса из контроллера в презентер;
- Перенес логику прогресса из контроллера в презентер;
- Перенес разбор ответа сервера из контроллера в презентер.

EntityVadim committed:
- Уменьшил код в ширину

EntityVadim committed:
- Создал конфигурацию в которой будут храниться все эти данные (AuthConfiguration);
- Создал хелпер (AuthHelper и AuthHelperProtocol);
- Сделал рефакторинг.

EntityVadim committed:
- Создал новый таргет для тестов (WebViewTests);
- Создал вьюконтроллер (WebViewTests);
- Создал презентер и соединил с вьюконтроллером (WebViewTests);
- Загрузил View и проверил значение viewDidLoadCalled (WebViewTests);
- Написал тест testPresenterCallsLoadRequest, чтобы проверить, вызывает ли презентер после вызова presenter.viewDidLoad() метод loadRequest вьюконтроллера;
- Тестируем необходимость скрытия прогресса;
- Написал тест testProgressHiddenWhenOne, чтобы проверить, что если значение прогресса равно единицы, то метод возвращает true.

EntityVadim committed:
- Создайте тест testAuthHelperAuthURL, чтобы проверить, что ссылка, полученная из authURL, содержит все необходимые компоненты;
- Написал тест testCodeFromURL, в котором нужно проверить, что AuthHelper корректно распознаёт код из ссылки.

EntityVadim committed:
- Переустановил библиотеки.

EntityVadim committed:
- Один тест не прошел (исправил это)

EntityVadim committed:
- Сделал рефакторинг ProfileViewController

EntityVadim committed:
- Сделал рефакторинг ImagesListViewController

EntityVadim committed:
- Пытаюсь покрыть тестами ProfileViewController

EntityVadim committed:
- Покрыл ProfileViewController тестами, но они не проходят, пытаюсь исправить эту проблему.

EntityVadim committed:
- Пытаюсь покрыть тестами ImagesListViewController

EntityVadim committed:
- Написал UI-тесты для приложения;
- Добавил идентификатор в webView;
- Мелкие корректировки по коду.

EntityVadim committed:
- Работаю с UI-тестами

EntityVadim committed:
- Сделал иконку приложения

EntityVadim committed:
- Удалил лишние иконки и мок-картинки

EntityVadim committed:
- Обновил UI-тесты;
- Доделал рефакторинг кода ProfileViewController;
- Перенес показ алерта о выходе из профиля в ProfileViewController.

EntityVadim committed:
- Вернул показ алерта обратно;
- Пишу unit-тесты.

EntityVadim committed:
- Доделал рефакторинг ImagesListViewController;
- Переписал юнит-тесты.

EntityVadim committed:
- Дописал юнит-тесты ProfileViewControllerTests;
- Мелкие изменения по коду ProfileViewController.

EntityVadim committed:
- Написал юнит-тесты для ImagesListViewController.

helecta commented on ImageFeed/Auth/AuthHelper.swift:
```diff
@@ -0,0 +1,46 @@
+import Foundation
+
+final class AuthHelper: AuthHelperProtocol {
+    let configuration: AuthConfiguration
```
лучше сделать приватным, доступ напрямую к конфигурации из вне не нужен, это внутреннее свойство, из которого AuthHelper делает нужные ссылки, запросы. Иначе у других разработчиков будет доступ к конфигурации и они могут начать придумывать свои методы для создания ссылок, которые будут не в AuthHelper

helecta commented on ImageFeed/Auth/WebViewPresenter.swift:
```diff
@@ -0,0 +1,42 @@
+final class WebViewPresenter: WebViewPresenterProtocol {
+    
+    weak var view: WebViewViewControllerProtocol?
+    var authHelper: AuthHelperProtocol
```
лучше сделать приватным и неизменяемым(private и let), т.к. инициализируешь с authHelper и это внутренее свойство

helecta commented on ImageFeed/Auth/WebViewPresenter.swift:
```diff
@@ -0,0 +1,42 @@
+
+final class WebViewPresenter: WebViewPresenterProtocol {
+    
+    weak var view: WebViewViewControllerProtocol?
```
view тоже можно вынести в init и сделать его let, не хорошо давать возможность изменять view извне, т.к презентер создаем для конкретного вью, и менять это вью по ходу жизни презентера не очень предсказуемый подход

helecta commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -20,42 +20,54 @@ final class WebViewViewController: UIViewController {
     
-    // MARK: - Update Progress
+    func setProgressValue(_ newValue: Float) {
+        progressView.progress = newValue
```
если использовать progressView.setProgress(…), то можно сделать заполнение прогресс бара более плавным.

helecta commented on ImageFeed/ImagesList/ImagesListPresenter.swift:
```diff
@@ -0,0 +1,61 @@
+
+final class ImagesListPresenter: ImagesListPresenterProtocol {
+    
+    weak var view: ImagesListViewControllerProtocol?
```
аналогично, view можно добавить в init и сделать его приватным и неизменяемым

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -9,19 +9,21 @@ import UIKit
-    private var photos: [Photo] = []
-    private var isLoadingNewPage = false
+    var presenter: ImagesListPresenterProtocol?
+    var photos: [Photo] = []
```
photos стоит перенести в презентер, т.к photos - это данные, контроллер не хранит в себе данные, он всегда получает их из презентера и все что он делает это отображает их нужным образом.

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -9,19 +9,21 @@ import UIKit
-    private let showSingleImageSegueIdentifier = "ShowSingleImage"
-    private var photos: [Photo] = []
-    private var isLoadingNewPage = false
+    var presenter: ImagesListPresenterProtocol?
```
можно сделать приватным, т.к у тебя есть метод configure, который устанавливает презентер. Иначе у тебя доступно два способа установки презентера, напрямую через свойство и через метод configure

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -47,16 +48,15 @@ final class ImagesListViewController: UIViewController {
         sender: Any?
     ) {
-        if segue.identifier == showSingleImageSegueIdentifier {
+        if segue.identifier == "ShowSingleImage" {
```
лучше строки "ShowSingleImage"  выносить в константы, так легче читать код и делать правки в будущем

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -69,46 +69,44 @@ final class ImagesListViewController: UIViewController {
-            } completion: { _ in
+            }, completion: { _ in
                 if diff != 0 {
                     self.tableView.reloadData()
```
В замыкание для избежания retain cycl'ов, необходимо передавать слабую ссылку на `self`

helecta replied:
А зачем в комплишене делать reload таблицы? Ты добавляешь с анимацией ячейки с помощью методов tableView.insertRows и tableView.deleteRows, дополнительно после этого обновлять таблицу не нужно

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -69,46 +69,44 @@ final class ImagesListViewController: UIViewController {
+        }
         let diff = newCount - oldCount
         photos = newPhotos
         if animated {
```
в этом методе надо оставить только то что связано с обновлением таблицы, а для обновления тебе нужны только indexPaths, их нужно получать из презентера

EntityVadim committed:
- Добавил тест в ImagesListViewControllerTests.

EntityVadim committed:
- Добавил еще два теста в ImagesListViewControllerTests.

EntityVadim committed:
- Исправил 2 замечания ревью;
- Организовывал код;
- Сделал некоторые свойства приватными.

EntityVadim committed:
- Просил комит

EntityVadim committed:
- Изменил видимость некоторых параметров;
- Сделать анимацию более плавной;
- Переписал тесты с учетом этих изменений.

EntityVadim committed:
- Удалил лишний код;
- "ShowSingleImage" вынес отдельно, для лучшей читаемости кода;
- Изменил код func updateTableView (обновление таблицы).

EntityVadim committed:
- Доделал рефакторинг в ImagesListViewController.

EntityVadim committed:
- Переписал тесты, с учетом прошлого рефакторинга.

EntityVadim committed:
- Удалил лишние расширения.
